xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,708 @@
+    if (!useProtocolV3 || listenerTemplate == null) {
+      listener.onNotServing(
+          Status.UNAVAILABLE.withDescription(
+              "Can only support xDS v3 with listener resource name template").asException());
+      initialStartFuture.set(null);
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+      return;
+    }
+    delegateBuilder.intercept(configApplyingInterceptor);
+    discoveryState = new DiscoveryState(listenerTemplate.replaceAll("%s", listenerAddress));
+  }
+
+  @Override
+  public Server shutdown() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalShutdown();
+        if (delegate != null) {
+          delegate.shutdown();
+        }
+      }
+    });
+    syncContext.drain();
+    return this;
+  }
+
+  @Override
+  public Server shutdownNow() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalShutdown();
+        if (delegate != null) {
+          delegate.shutdownNow();
+        }
+      }
+    });
+    syncContext.drain();
+    return this;
+  }
+
+  // Must run in SynchronizationContext
+  private void internalShutdown() {
+    logger.log(Level.FINER, "Shutting down XdsServerWrapper");
+    if (discoveryState != null) {
+      discoveryState.shutdown();
+    }
+    if (xdsClient != null) {
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+    }
+    if (restartTimer != null) {
+      restartTimer.cancel();
+    }
+    isServing = false;
+    internalTerminationLatch.countDown();
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return shutdown.get();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    if (internalTerminationLatch.getCount() != 0) {
+      return false;
+    }
+    if (delegate != null) {
+      return delegate.isTerminated();
+    }
+    return true;
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    if (!internalTerminationLatch.await(timeout, unit)) {
+      return false;
+    }
+    if (delegate != null) {
+      delegate.awaitTermination();
+    }
+    return true;
+  }
+
+  @Override
+  public void awaitTermination() throws InterruptedException {
+    internalTerminationLatch.await();
+    if (delegate != null) {
+      delegate.awaitTermination();
+    }
+  }
+
+  @Override
+  public int getPort() {
+    if (delegate != null) {
+      return delegate.getPort();
+    }
+    return -1;
+  }
+
+  @Override
+  public List<? extends SocketAddress> getListenSockets() {
+    if (delegate != null) {
+      return delegate.getListenSockets();
+    }
+    return Collections.emptyList();
+  }
+
+  @Override
+  public List<ServerServiceDefinition> getServices() {
+    if (delegate != null) {
```

Comment:
This sort of logic seems just wrong. What was wrong with the previous behavior where it continued using the old server until it was replaced?

Reply:
Hm, it seems the same behavior with the previous behaviour (I believe you mean `ServerWrapperForXds`). It keeps using the old server until it is replaced.
The difference is that the previous `delegate` is built immediately in constructor, so it is never null. But now I want to wait until I install the `ConfigApplyingInterceptor` so it might be null.
Note that before and after, the volatile `delegate` might be shutdown and then rebuilt, because we can not restart the same server once shutdown.

Reply:
Hmm... you are right it is the same in that it uses the old server until it is replaced. I saw the `== null` logic and assumed that was what it was for. But it is at least broken before the server is started. Why not add `ConfigApplyingInterceptor` in the constructor?

Reply:
> The difference is that the previous delegate is built immediately in constructor, so it is never null. But now I want to wait until I install the ConfigApplyingInterceptor so it might be null.
> Why not add ConfigApplyingInterceptor in the constructor?

Yes we can put `ConfigApplyingInterceptor` in constructor, I just felt putting in `internalStart()` is more clear because `internalStart()` does all the preparation work. 
Also `==null` logic is negotiable.  We can build immediately in constructor, making it always `!=null`. 
This also means calling some `get()` APIs  would throw exception due to delegate not `started`, and getting meaningful results of for `getService()`, etc., and also we may shutdown() delegate before start due to errors. 
This would actually be simpler but I was wondering is it right to expose those APIs before we validate any xds configurations?

Reply:
> Yes we can put ConfigApplyingInterceptor in constructor, I just felt putting in internalStart() is more clear because internalStart() does all the preparation work.

I think it optimizes one thing at the expense of another. The bigger problem for this PR though is that it changes a fundamental behavior of the class and so requires a complete re-review of the code. There is some benefit of not putting it in `internalStart()` because internalStart is really about lazy resource acquisition and there's benefit in it not being cluttered to make it easier to audit object lifetimes.

> This would actually be simpler but I was wondering is it right to expose those APIs before we validate any xds configurations?

We'd have to look at each method in particular. While it may be a useful thing to think about, that seems independent from the xds routing and config work; the previous code still had to load config and if the previous code was wrong we could fix it before or after the xds routing and config changes.